src/firefly/js/ui/NaifidPanelWorker.js

```diff
@@ -0,0 +1,142 @@
+/*
+ * License information at https://github.com/Caltech-IPAC/firefly/blob/master/License.txt
+ */
+
+//-----------------
+import {getRootPath} from '../util/BrowserUtil.js';
+import {fetchUrl} from '../util/WebUtil.js';
+
+
+export {formatPosForTextField} from '../data/form/PositionFieldDef.js';
+
+
+
+function makeResolverPromise(objName) {
+    let ignoreSearchResults= null;
+    let aborted= false;
+
+    const aborter= function() {
+        aborted= true;
+        if (ignoreSearchResults!==null) ignoreSearchResults();
+    };
+
+    const workerPromise= new Promise(
+        function(resolve, reject) {
+            setTimeout( ()=> {
+                if (aborted) {
+                    reject();
+                }
+                else {
+                    const {p, rejectFunc}= makeSearchPromise(objName);
+                    ignoreSearchResults= rejectFunc;
+                    resolve(p);
+                }
+            }, 200);
+        }
+    );
+
+    return {p:workerPromise, aborter};
+}
+
+
+
+function makeSearchPromise(objName) {
+    let rejectFunc= null;
+    const url= `${getRootPath()}sticky/CmdSrv?objName=${objName}&cmd=CmdResolveNaifid`;
+    const searchPromise= new Promise(
+        function(resolve, reject) {
+            let fetchOptions = {};
+            // AbortController might not be available in older browsers
+            if (typeof AbortController !== 'undefined') {
+                // fetch will be aborted after timeout
+                const fetchTimeoutMs = 7000;
+                const controller = new AbortController();
+                const signal = controller.signal;
+                setTimeout(() => {
+                    controller.abort();
+                }, fetchTimeoutMs);
+                fetchOptions = {signal};
+            }
+
+            fetchUrl(url, fetchOptions).then( (response) => {
+                response.json().then((value) => {
+                    resolve(value);
+                });
+            }).catch( (error) => {
+                return reject(error);
+            });
+        });
+
+    const abortPromise= new Promise(function(resolve,reject) {
+        rejectFunc= reject;
+    });
+    return {p:Promise.race([searchPromise,abortPromise]), rejectFunc};
+}
+
+
+export function resolveNaifidObj(object){
+    let result = resolveObject(object);
+    return result;
+}
+
+
+
+
+function resolveObject(objName) {
+    if (!objName) {
+        return {
+            showHelp: true,
+            valid : true,
+            feedback: ''
+        };
+    }
+
+    let {p}= makeResolverPromise(objName);
+    p= p.then( (results) =>
```

Comment:
This is misleading... why re-assign `p` to a `then` declaration?
It should be written,  `p.then( (results) =>`  without `p = `.

Reply:
OK, the problem is that you want to return the then promise not p. So what @deannaji is doing is correct, but confusing. It should just be `return p.then...`

But if you do then you will needs to modify `getSuggestions` so it is expect a `promise` from `resolveNaifidObj` and not an `object`.
 
NaifidPanel.jsx, `getSuggestions`
```
            if (!rval) return undefined;
            return rval.then((response)=>{
```

I think you copied some code from target panel where it was return not only a promise in the object but also a function to abort the promise. Since you are only returning a promise there is no need to wrap it in an object.

Reply:
Done, thanks!!